ocaml/gencert/validation.ml

```diff
@@ -51,58 +51,197 @@ let validate_private_key pkcs8_private_key =
                ] )
          else (
            D.info {|Failed to validate private key because "%s"|} err_msg ;
-           `Msg (server_certificate_key_invalid, [])))
+           `Msg (server_certificate_key_invalid, [])
+         ))
   >>= ensure_key_length
 
-let validate_certificate kind pem now private_key =
-  let ensure_keys_match private_key certificate =
-    let public_key = X509.Certificate.public_key certificate in
-    match (public_key, private_key) with
-    | `RSA pub, `RSA priv when pub = Mirage_crypto_pk.Rsa.pub_of_priv priv ->
-        Ok certificate
-    | _ ->
-        Error (`Msg (server_certificate_key_mismatch, []))
-  in
+let has_valid_time time certificate =
+  let not_before, not_after = X509.Certificate.validity certificate in
+  Ptime.(is_later ~than:not_before time && is_earlier ~than:not_after time)
+
+let load_server_certificate pem =
```

Comment:
Similarly to my comment about anchors, I think it would make sense to perform the cert validation in this constructor (unless there's a reason for keeping expired certs around?)

Reply:
This function does not exist and its functionality is contained in `validate_server_certificate`